### ISSUE TITLE
Load balancer convergence: draining state

### ISSUE THREAD

**Issue by cyli** [rackerlabs/otter#715]
When we scale down or otherwise delete a healthy server, it should finish serving the requests it is currently serving before being removed from the load balancer and deleted.  (This probably does not apply to servers that have gone into error state from active, because they're likely not serving traffic anyway).  Will be referring to this as "the draining feature" from here on out.

Note that this may be an optional feature - if the user doesn't care about draining, although that's unlikely, the server can just be removed from all load balancers and deleted.

Assuming the user wants the draining feature, the workfow for deleting a server would be:
1. Change the node condition to draining for all of the load balancers that the server is on.
2. When there are no more connections to the node from the load balancer (this statistic will be added by the load balancer team in the future), we can remove the node from the load balancer.
3. After X seconds (user specified parameter on the group, so no per-load balancer timeouts), we remove the node from the load balancer anyway whether or not there are still connections.
4. We delete the server.
### Summary of tasks:

These are expounded on further down, but here is the short version of proposed changes without any justifications or explanations:
- [x] **code changes/additions**
  - [x] flag servers via metadata as being in 'deleting-by-autoscale' state
  - [x] exclude 'deleting-by-autoscale' servers from active servers
  - [x] _for each 'deleting-by-autoscale' server, for each load balancer it is attached to:_
    - [x] _put node into draining if it is enabled_
    - [x] _if node is draining_:
      - [x] _check per-node load balancer stats_
      - [x] _remove server from load balancer once all the connections have been terminated, or the draining timeout expires_
  - [x] _for each 'deleting-by-autoscale' server, delete server once it is no longer attached to any load balancers_
- [x] **request node-last-updated info per node from the load balancer team** so we can know whether the timeout has passed (workarounds below, if this is infeasible)
- [X] **answer question: if we move a server to a different load balancer, do we still do the draining rigamarole?**  If so, more features/changes may need to be added - **ANSWER**: Not now
### Explanations for the task items above:
- **How to count active servers**:
  
  We can't just go by the Nova status, because it may have been put into draining by autoscale and in the process of being deleted.  We can't just filter out all servers that are in draining in all their load balancers (maybe for some reason the user specified in the launch config that all servers start out in draining, maybe the user changed it to draining and autoscale should converge it back to enabled).
  
  We need to flag it a server in deleting by autoscale standards.  Possibly we can just filter out all active servers with a particular metadata key.  
- **Converging deleting servers**:
  
  The steps above (_in italic_) for converging the deleting servers, once they can be identified, seem fairly self-explanatory.  The only addendum is that the desired state (as per the server metadata) of a deleting server can probably be ignored.
- **Keeping track of the draining timeout** - How do we know how long a node has been in draining?  How do we know that a timeout has expired and we can remove the node from the load balancer?
  **UPDATE** - We can depend on load balancers for this information, like we depend on the created time in Nova details to determine when to timeout a server in building state.  We can check the user change events in `/loadbalancers/{lb_id}/nodes/{node_id}.atom`, which gives user change events from latest to earliest.
  1. <del>Server update time - this is the update time of the server, which Nova gives us.  This is because we assume the last change made to the server was by us, flagging it as to-be-deleted by autoscale.  However, it may take a while to actually change the server to draining on every single load balancer, since load balancers are immutable for some time right after a change.  If we go with this timestamp, it is very likely to be inaccurate for at least one load balancer. **UPDATE**: this isn't really dependable - see @manishtomar's comments below </del>
  2. <del> Store timestamp in metadata - every time we successfully put a server into draining condition on a load balancer, we store the timestamp for that load balancer as a different metadata key.  This may get more complicated since there can be an arbitrary number of keys (there are an arbitrary number of loadbalancer-port mappings per server - if we need to put nodes into draining to just move a load balancer, then the timestamp also needs to be deleted, we may need to mutate desired LB state, etc).</del>
  3. <del>Depend on load balancers for this information - we depend on the created time in Nova details to determine when to timeout a server in building state.  There is no corresponding thing in the load balancers API.  There is [an endpoint to get node events](http://docs.rackspace.com/loadbalancers/api/v1.0/clb-devguide/content/GET_listNodeEvents_v1.0__account__loadbalancers__loadBalancerId__nodes_events_Nodes-d1e2173.html) on the CLB API, but some testing indicates that it does not include user-initiated changes to the condition.  Just internal changes to the node status.  (E.g. if the load balancer thinks the node is offline). </del>
     
     
     <del>We should try to ask the load balancer team for this information, since this should live in their system.  This would be my preference, since this would be the time the load balancer changed the node state, as opposed to the time a response from load balancers came back to otter.</del>
     
     
     <del>However, then this feature would be blocked on their providing this information. </del>
  
  <del>We can also try to keep draining time in the DB, but I did not include that here as an option because then the true state of the world would be a combination of data in the DB and data in Nova, and we should probably just have a single source of truth.</del>

<a name="draining_on_moving_lb" />    
### Other question

Currently we are removing a server from all load balancers it should not be attached to.  Should the node be put into draining before it is removed?

If the user changes the load balancer in the launch config, and we converge to the new desired state, should the node be put into draining on the old load balancers first?

**UPDATE - 10/23/2014 afternoon** With per-group timeouts, this is actually pretty easy to support.

<del>**UPDATE - 10/23/2014 morning** If we do this, how long do we put the server into draining?  </del>

1.  <del>User has changed the launch config to reflect a new set of load balancers.  We no longer store desired load balancers in the metadata, so we've lost the information on how long the servers should be in draining on the old load balancers.</del>
2.  <del>User manually put a server into a different load balancer - currently we remove it.  There never was any draining timeout for this load balancer - do we put it into draining, and for how long?</del>

<del>One possibility is to use per-node load balancer metadata, but load balancer metadata has a few drawbacks:</del>

- <del>it cannot be included when adding the node - it must be added after in a separate call</del>
- <del>it does not appear when listing nodes - node details must be fetched</del>
- <del>metadata is not a dictionary - it's an array of dictionaries that must be iterated over - looks like:</del>
  <del>
        {
        "metadata": [
          {
            "id":"1",
            "key":"color",
            "value":"red"
          },
          {
            "id":"2",
            "key":"label",
            "value":"web-load-balancer"
          }
        ]
      }
  </del>
- <del>deleting metadata must be done with that metadata item's ID</del>
- <del>each update to metadata seems to put the load balancer in 'pending update' state during which no other changes can be made to the load balancer</del>

<del>So using per-node load balancer metadata means a lot more calls to the load balancer to get and set the relevant information, and it may be difficult to enforce per-node metadata consistency</del>

<del>Another possibility is to just have general timeout value that can be overridden by the desired load balancer configuration.</del>

<del>A third possibility is to not have draining timeout per load balancer, but just 1 draining timeout for the whole group, which would apply to EVERY load balancer.</del>
